base/linalg/diagonal.jl

```diff
@@ -180,8 +180,12 @@ end
 (\)(D::Diagonal, b::AbstractVector) = reshape(scale(1 ./ D.diag, reshape(b, length(b), 1)), length(b))
 (\)(Da::Diagonal, Db::Diagonal) = Diagonal(Db.diag ./ Da.diag)
 
+invsimilar{T}(d::Vector{T}) = similar(d, T)
+invsimilar{T<:Integer}(d::Vector{T}) = similar(d, Float64)
+invsimilar{T<:Integer}(d::Vector{Complex{T}}) = similar(d, Complex{Float64})
+
 function inv{T}(D::Diagonal{T})
-    Di = similar(D.diag)
+    Di = invsimilar(D.diag)
```

Comment:
It is better to use a generic promotion here. Something like
```jl
S = isleaftype(T) ? typeof(inv(zero(T))) : typeof(inv(D.diag[1]))
Di = similar(D.diag, S)
```
That would also work for element types that are not integers, but get promoted under inversion.